### ISSUE TITLE
Autocorrect tooltips don't appear on iOS

### ISSUE THREAD

**Issue by jwmcglynn** [flutter/flutter#12920]
## Steps to Reproduce

When using a native iOS app without text prediction enabled, there is a pop-up that indicates when autocorrections are going to occur, and when backspacing on a correction a pop-up appears that gives alternative choices for the word.

In flutter, the autocorrections happen silently and require backspacing into the word to make the correction.  See the gif below for a comparison of the behavior.

Flutter should either integrate with the iOS autocorrection system to provide the popups or emulate them.  This is a pretty big usability concern which makes it more difficult to enter text because the corrections are harder to notice and harder to fix.

![text_correction](https://user-images.githubusercontent.com/103489/32534122-0182048a-c409-11e7-8879-606857aff89e.gif)

## Logs

N/A, this is a design issue.

## Flutter Doctor

```
$ flutter doctor
Building flutter tool...
[✓] Flutter (on Mac OS X 10.12.6 16G29, locale en-US, channel master)
    • Flutter at /Users/jwm/Projects/flutter
    • Framework revision 337150308c (2 days ago), 2017-11-05 19:39:49 -0800
    • Engine revision e059cc0258
    • Tools Dart version 1.25.0-dev.11.0

[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.io/setup/#android-setup for detailed instructions).
      If Android SDK has been installed to a custom location, set $ANDROID_HOME to that location.

[✓] iOS toolchain - develop for iOS devices (Xcode 9.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.0.1, Build version 9A1004
    • ios-deploy 1.9.2
    • CocoaPods version 1.3.1

[✗] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.io/setup/#android-setup for detailed instructions).

[✓] IntelliJ IDEA Community Edition (version 2017.2.5)
    • Flutter plugin version 18.2
    • Dart plugin version 172.4343.25

[✓] Connected devices
    • iPhone 7 Plus • DF665BBF-F7D6-4B75-8FCA-E092AD4AD6DC • ios • iOS 11.0 (simulator)
```

**Comment by eseidelGoogle** [comment]
MD doesn't seem to talk about this, so this is likely going to be cupertino specific.
https://material.io/guidelines/components/text-fields.html
Thoughts @cbracken @xster?

**Comment by lukepighetti** [comment]
@eseidelGoogle iOS doesn't have autocorrect on the keyboard like Android does, so this is not MD vs Cupertino, but more like we just need it in all apps on iOS.

**Comment by xster** [comment]
Thanks for the feedback. We're addressing a few non-cupertino iOS text entry issues at the moment like https://github.com/flutter/flutter/issues/9507, https://github.com/flutter/flutter/issues/20693 and https://github.com/flutter/flutter/issues/9288. Hopefully we'll have the bandwidth to address this one too after.

**Comment by lukepighetti** [comment]
You know what's kind of embarrassing? I was using my phone today and autocorrect was coming up on the keyboard. Literally never seen or noticed it before. So my apologies for my incorrect statement earlier. I'm not sure when that changed or when it displays. This was on iOS 10 btw.

**Comment by Ethras** [comment]
What's the status on this?

**Comment by zoechi** [comment]
@Ethras 
Using "add reaction" on the initial comment would increase priority while `+1` comments are rather pointless and cause lots of people to get notified for no reason.
To get notified yourself use the [Subscribe] button to the top right instead.

**Comment by jslavitz** [comment]
This is currently in progress!

**Comment by jakub-bacic** [comment]
Any ETA regarding the fix?

**Comment by xster** [comment]
There's a solution in work but our work allotment was reshuffled recently. I would realistically expect this to be done closer to June.

**Comment by LongCatIsLooong** [comment]
According to UIKit documentation, some of the behaviors described above come from [spellchecking](https://developer.apple.com/documentation/uikit/uitextinputtraits/1624461-spellcheckingtype) rather than [autocorrection](https://developer.apple.com/documentation/uikit/uitextinputtraits/1624453-autocorrectiontype) (Although when I tested on iOS 12 & 13 if autocorrection is on turning off spellchecking only hides the red underline).

Created #34688.

**Comment by dvxf** [comment]
Please note that the popup shown in the original post is only shown if "Predictive" is turned to OFF in the General->Keyboard settings. By turning off this setting, you don't get 3 auto-correct choices at the top of the keyboard, but just one right next to the word you're typing in, exactly as shown in the original post, and a cross to dismiss the correction.

It was the default behaviour of iOS until the top-of-the-keyboard thing was included, so personally I keep this setting off on my phone as I'm used to it. My guess it that the percentage of users with this setting off is small as it's not the default behaviour anymore, but I'm surely not alone.

Unfortunately, Flutter textfields are almost unusable in this configuration, as you have no idea what word will be replaced by the auto-correct.

**Comment by lukepighetti** [comment]
Thanks for the explanation @dvxf, could never figure out why others had the top-of-the-keyboard thing and I don't. Flutter should be designed to work without the top-of-the-keyboard thing without a doubt.

**Comment by LongCatIsLooong** [comment]
Autocorrected words are now highlighted on iOS. Closing.

**Comment by jakub-bacic** [comment]
@dvxf 
Just a little comment from my side - it's also a default behavior when... there's no prediction implemented for the given language (e.g. Polish).

Anyway, I'm glad it's fixed now 🎉

**Comment by dvxf** [comment]
@LongCatIsLooong 
Thank you, it's definitely an improvement that autocorrected words are now highlighted on iOS! However, it only fixes half of the issue: as far as I can tell, the autocorrect tooltip still doesn't appear on iOS.

So it improves the behaviour when Predictive Keyboard is turned on, but not when Predictive Keyboard is turned off (or in Polish as @jakub-bacic mentioned).

As I mentioned before, unfortunately Flutter textfields are almost unusable in this configuration, as you have no idea what word will be put by the autocorrect in place of the highlighted word.

**Comment by lukepighetti** [comment]
@LongCatIsLooong it appears that there was a misunderstanding about this issue. This issue is about tooltips, not squiggly line highlighting.

**Comment by lukepighetti** [comment]
@dvxf do you happen to know if there is any way to turn on the Predictive Keyboard in one app? We could shim this issue with that for the time being if it's possible.

**Comment by LongCatIsLooong** [comment]
Oh my apologies. I didn't realize the tooltip can still appear in some cases, reopening.
It seems the tooltip will be difficult to implement at the moment: we can't draw the tooltip in flutter, and currently there doesn't seem to be a good way to return the accurate `CGRect` in `func firstRect(for range: UITextRange) -> CGRect` in the text input plugin. 

Related: https://github.com/flutter/flutter/issues/9344

**Comment by lukepighetti** [comment]
Even if we relocated it to be above the text line instead of below, that would be fine.

**Comment by xster** [comment]
@LongCatIsLooong is the remaining issue still a customer critical?

**Comment by LongCatIsLooong** [comment]
@xster I would think it's still blocking at least for apps that target Polish users? 

@lukepighetti I think anchoring the tooltip to the text input's bounds would confuse the users, especially when the text field is multiline.

**Comment by lukepighetti** [comment]
I'm confused by the lack of priority on this issue. I am an English iOS user and I have no controls for autocorrect in any of my Flutter apps on iOS.

**Comment by Hixie** [comment]
I recommend filing separate bugs for the remaining work so that we can track each specific item separately. Having one issue track a large number of separate work items doesn't lend itself well to clear prioritisation.

For the new bugs, I recommend marking them customer critical, etc, as per: https://github.com/flutter/flutter/wiki/Issue-hygiene#prioritization

**Comment by Hixie** [comment]
(Please list the new bugs filed here, so that people who are following this one can follow up with the relevant bugs.)

**Comment by lukepighetti** [comment]
This issue tracks the remaining work precisely, it was just misunderstood.

**Comment by Hixie** [comment]
Got it, thanks. Sorry, I misread the bug earlier.

**Comment by Hixie** [comment]
@LongCatIsLooong Any updates?

**Comment by LongCatIsLooong** [comment]
@Hixie Currently working on 55613. I will start looking into this more closely after that.

**Comment by LongCatIsLooong** [comment]
I spoke with @GaryQian and @HansMuller about this.

Proposed solutions so far generally fall into these 2 categories:

- Blocking execution

Delegate the [`func firstRect(for range: UITextRange) -> CGRect`](https://developer.apple.com/documentation/uikit/uitextinput/1614570-firstrect) computation to the framework using platform channels, block the platform thread until the result comes back. See https://github.com/flutter/engine/pull/4358 for potential problems with blocking the platform thread.

- Asynchronous prefetching 

Keep a copy of `Paragraph` data (or whatever information it takes) in the iOS text input plugin, and keep it up to date with the framework using platform channels. Implement [`func firstRect(for range: UITextRange) -> CGRect`](https://developer.apple.com/documentation/uikit/uitextinput/1614570-firstrect) using the engine copy of information. 
  
This introduces a race condition: when text is entered via the software keyboard, the iOS text input plugin sends the update back to the framework (which in turn triggers an update in `Paragraph` data that needs to be synced back to the engine side), and at the same time the keyboard calls [`func firstRect(for range: UITextRange) -> CGRect`](https://developer.apple.com/documentation/uikit/uitextinput/1614570-firstrect). There's no guarantee the updated `Paragraph` data will arrive in the engine before [`func firstRect(for range: UITextRange) -> CGRect`](https://developer.apple.com/documentation/uikit/uitextinput/1614570-firstrect) returns. And there's no way to tell iOS to update the rect should the `Paragraph` data arrives late.

**Comment by matehat** [comment]
@LongCatIsLooong I don't know if your comments was meant to indicate that this issue was blocked or reconsidered, but I wanted to add some weight to this.

We're building a secure messaging app for healthcare, and 50% of the user experience is around typing. Currently on iOS, unless you set your iPhone to have the 3 autosuggestions above the keyboard, the typing experience is simply broken. Some replacements that are usually shown using the white tooltip before being applied do not appear, creating confusion and straight inability to prevent said replacement.

Below, on the left is a reference working example, using Notion (probably react-native or something like this), and on the right, our app using Flutter.

![2020-07-14 06 44 40](https://user-images.githubusercontent.com/1283/87417143-c0003d80-c59d-11ea-8768-05770cc3b06f.gif)

I want to emphasize this: the experience is _currently broken_. I cannot tell our users to go in the settings and change their keyboard specifically for our app. It _is_ high priority, not a minor visual glitch.

**Comment by lukepighetti** [comment]
What's especially alarming to me is the lack of priority on the team side and the lack of severity on the community side. We have started recommending React Native for any heavily text based apps that are going on the iOS App Store because of this broken experience.

**Comment by matehat** [comment]
@lukepighetti that's understandable, and unfortunate since Flutter is such a great framework. 

On our end, being a bootstrapped startup with limited resources we simply can't afford to switch to another tech at this point. We made the leap and invested heavily in Flutter. I'm just really hoping the team fixes it soon, since our app really looks clumsy and buggy because of our tech choices.

/cc @Hixie @LongCatIsLooong please, please increase the bug priority. It's a pretty big deal.

**Comment by lukepighetti** [comment]
Maybe someone from the flutter team can give us a high level strategy for implementing it. I don't have the bandwidth to work on this feature ATM but if they already have an idea of how to implement then someone someone may have the time to achieve it.

**Comment by LongCatIsLooong** [comment]
@matehat the gif on the right confuses me a bit, at least the background highlight should have shown up when the text was autocorrected.

For autocorrect as previously mentioned in `func firstRect(for range: UITextRange) -> CGRect` a correct `CGRect` needs to be returned for iOS to be able to correctly position the text background highlight and the tooltip. Currently we don't have an easy way to get the correct `CGRect` synchronously, so we have 2 options:

1. Return `CGRectZero` to prevent iOS from drawing the highlight and the tooltip, and send the `UITextRange` provided in the call to the framework to let Flutter draw the highlight. This is the current approach, it should work well for the highlight but we lose the tooltip.

2. Try our best to predict the correct `Rect` and let iOS draw the highlight & the tooltip. I made a POC that seems to work well enough for #9344, but in the case of autocorrect, the entire `Rect` needs to be on point (instead of just the leading edge or the trailing edge), so the background highlight rect iOS shows indicates the correct range of the text that will be autocorrected.

**Comment by matehat** [comment]
@LongCatIsLooong I'm using the latest stable:

```
Flutter 1.17.5 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 8af6b2f038 (2 weeks ago) • 2020-06-30 12:53:55 -0700
Engine • revision ee76268252
Tools • Dart 2.8.4
```

Here are the exact keyboard settings: 

![PNG image](https://user-images.githubusercontent.com/1283/87469546-72a9bd80-c5e9-11ea-894d-5d2763dc3074.png)

The whole puzzle is also because we're trying to do all of this with _asynchronous_ callbacks between the host platform and the framework if I understand correctly, right?

**Comment by LongCatIsLooong** [comment]
> The whole puzzle is also because we're trying to do all of this with asynchronous callbacks between the host platform and the framework if I understand correctly, right?
Yes, we'd like to avoid doing synchronous communication: flutter.dev/go/flutter-platform-idling. 

I noticed in the second gif nothing got highlighted when autocorrect took place. That looks like a different bug that needs fixing? In my test flutter app I can make the text highlight show up with the same keyboard configuration, by following these steps:
1. turn predictive off
2. in the flutter app,  switch keyboard to French
3. type "ca va bien" ("ca" is highlighted, and if I continue to type it's autocorrected)

**Comment by lukepighetti** [comment]
What do you mean by highlighted? The squiggly line below text?

**Comment by LongCatIsLooong** [comment]
@lukepighetti  No I was referring to the surrounding light gray correction rectangle. Maybe there's a way to change its color to transparent to make it invisible, so we don't have to worry about having the precise rectangle.

**Comment by lukepighetti** [comment]
I have literally never seen the light gray correction rectangle in a Flutter app on iOS. Can you show us a screenshot of what you're talking about on Flutter?

**Comment by LongCatIsLooong** [comment]
Sorry I forgot it's light blue by default:
<img width="133" alt="Screen Shot 2020-07-15 at 1 53 45 PM" src="https://user-images.githubusercontent.com/31859944/87595009-b90f2300-c6a2-11ea-8cf4-eb5fe136d615.png">

**Comment by lukepighetti** [comment]
I thought that was only shown when selecting text?

**Comment by lukepighetti** [comment]
Gotcha. Yeah, this issue is about the autocorrect suggestions that appear next to the iOS cursor while you're typing regardless of the styling of the app.

Currently it works as expected if you have `Predictive` setting turned on in `General => Keyboards`.

However, if that's turned off these tooltips do not appear when typing.

<img width="384" alt="Screen Shot 2020-07-16 at 7 02 19 AM" src="https://user-images.githubusercontent.com/666539/87663814-667d4780-c732-11ea-9305-64d3a2209980.png">

There are two solutions in my mind.

1) Find a way to force `Predictive` setting whenever in a Flutter app.
2) Implement the autocorrect tooltips for when `Predictive` is off.

**Comment by matehat** [comment]
It doesn't seem like option 2 is possible if I look at all the documentation and questions around predictive UI.

I think showing the tooltip is more important than showing the exact highlighted region, since the tooltip gives you the ability to accept or refuse a correction. 

Is there a way to show the tooltip and position it approximately right?

**Comment by matehat** [comment]
Sorry I just saw this part of @LongCatIsLooong comment:

> It seems the same Rect returned is used for both the highlight and the tooltip, and returning a Rect that has 0 width (or height) would also prevent both the tooltip and the highlight from showing up, so we can't just tell iOS to hide the highlight and show the tooltip.

Maybe a minimal but non-zero width, and a height deducted from the text's known line height would do?

**Comment by LongCatIsLooong** [comment]
I found an approach that seems to work well for https://github.com/flutter/flutter/issues/9344, save for a few edge cases. I'll see if that works for the autocorrect tooltip too. The highlighted region must accurately indicate the range of the autocorrect otherwise it can be misleading. Maybe we can override `addSubview` to prevent the highlight from being added.

**Comment by lukepighetti** [comment]
@matehat did you end up finding any immediate relief? I just had a UX testing session with someone entering tags and they were having a really hard time getting by without tooltips.

Do we know if there's a way to at least detect if predictive keyboard is enabled? We could then show them a message suggesting they turn it on.

What I've noticed is that people who are long time iPhone users have this feature disabled by default and don't even know it exists. If you're using an iPhone on an account that isn't many years old then you might not realize that iOS had this tooltip feature.

**Comment by lukepighetti** [comment]
There's another option. If we can get autocorrect options programmatically we can turn off autocorrect on the field and display our own predictive text options in Flutter. That would also allow us to add in our own actions which is a value add over the native controls (afaik).

There's also this method for detecting if Predictive is enabled. https://stackoverflow.com/questions/25993949/check-if-predictive-text-is-enabled

**Comment by henryleunghk** [comment]
I encountered the same situation like you guys. I have been searching for a period of time looking for a solution to have same text input experience as other iOS apps built with either native Swift/Objective-C/React-Native. Finally, I resorted to writing my own package to export `UITextView` from iOS for use in Flutter:
https://pub.dev/packages/flutter_native_text_input

I hope you guys find it useful. Please feel free to report any issues or create PRs for any missing features.
Cheers🍻

**Comment by joniaiuser** [comment]
Any chance if the iOS experience will be incorporated into Flutter natively?

**Comment by joniaiuser** [comment]
@LongCatIsLooong Any updates for this?

**Comment by FloLecoeuche** [comment]
@LongCatIsLooong Can you add this feature to the Flutter roadmap ?

**Comment by LongCatIsLooong** [comment]
@FloLecoeuche sorry I don't think I can provide a good time estimate, I'm running out of ideas of how the rest of iOS autocorrect can be implemented without introducing obvious visual glitches. Unassigning myself for now but I'll keep thinking about this, and would be happy to discuss new ideas.

**Comment by lukepighetti** [comment]
So this is a `wontfix`?

**Comment by HansMuller** [comment]
This isn't a "wontfix", we're just trying to be transparent about our efforts to fix it so far.

**Comment by tomaschyly** [comment]
As this is critical feature for many apps and there is still no official fix, I have decided to create my own solution.

My package supports null-safety and allows theming. You don't have to use my theme, the widget works standalone.
It is hybrid solution where there are theme-able Flutter widgets wrapping UITextField and UITextView for just text, which can be customized too.

Maybe Flutter team should consider similar solution.

https://pub.dev/packages/tch_common_widgets

**Comment by henryleunghk** [comment]
> I encountered the same situation like you guys. I have been searching for a period of time looking for a solution to have same text input experience as other iOS apps built with either native Swift/Objective-C/React-Native. Finally, I resorted to writing my own package to export `UITextView` from iOS for use in Flutter: https://pub.dev/packages/flutter_native_text_input
> 
> I hope you guys find it useful. Please feel free to report any issues or create PRs for any missing features. Cheers🍻

Android support is added in v2😀

**Comment by Hacklone** [comment]
I fail to understand how this is only P3. This is a very serious issue with flutter as a core typing functionality of iOS is **broken**.

**Comment by HongliYu** [comment]
There is no perfect solution yet, so I bring this bridging demo to help the community enrich the knowledge. @Hacklone 
[ios_native_text_field_bridging](https://github.com/HongliYu/ios_native_text_field_bridging)

**Comment by justinmc** [comment]
CC @cyanglaz @camsim99 I just noticed this older issue and wanted to put it on your radar since it partially has to do with spellcheck on iOS.

**Comment by bryanoltman** [comment]
+1 on this – this is probably the fit-and-finish issue I notice most on Flutter iOS apps. Also, I noticed this was tagged "cupertino", but it is not only a cupertino issue (it happens with Material text fields as well).

**Comment by jtkeyva** [comment]
+1

**Comment by joniaiuser** [comment]
Should be raised to P1

**Comment by justinmc** [comment]
I think there are two main features to implement here:

### 1. Spellcheck

A red underline appears under misspelled words.  The normal context menu shows suggestions that can be tapped to replace.

![Screenshot from 2022-10-24 13-47-16](https://user-images.githubusercontent.com/389558/197626524-e0ee10aa-71f1-4f6e-b126-4d092038314c.png)
![Screenshot from 2022-10-24 13-38-12](https://user-images.githubusercontent.com/389558/197626527-72fba6a2-a52a-4968-b77d-77dcc487140d.png)

This is currently in progress for Android and iOS:

  * [x] Android engine work: https://github.com/flutter/engine/pull/32941
  * [x] iOS engine work: https://github.com/flutter/engine/pull/30858
  * [ ] Context menu refactor: https://github.com/flutter/flutter/pull/107193
  * [ ] Android framework: in progress by @camsim99 
  * [ ] iOS framework: in progress by @camsim99 (after Android)

### 2. Autocorrect

A blue highlight appears on a word available for autocorrect as the user types, along with a menu showing replacement options.

![Screenshot from 2022-10-24 13-37-51](https://user-images.githubusercontent.com/389558/197626529-2fba7930-fb49-48b0-bc43-5ca85d2557d5.png)

This has been investigated by @LongCatIsLooong in the comments above and is up next for implementation.

**Comment by alxmoroz** [comment]
5+ years still open...
Maybe it's all from the evil one and it's time to return to native development?

**Comment by knopp** [comment]
@LongCatIsLooong, there is already a precedent for blocking call from engine to UI thread in `FlutterKeyboardManager.mm`. It runs nested CFRunLoop in `kMessageLoopCFRunLoopMode` until it gets the response.

Looking at the code I am wondering a bit what happens if the UI isolate gets terminated when within the inner loop. This is a bit underdocumented and I'm not 100% sure it is handled correctly.

Arguably this is playing with fire a little bit, but in this case it should be perfectly safe since the UI thread action is completely synchronous and UI thread can not do blocking calls to other threads.

We do similar thing in `super_drag_and_drop` to get items for `UIDragInteractionDelegate` and `UIContextMenuInteractionDelegate` (though using FFI instead platform channel in order to make sure not to deadlock when isolate is terminated before platform code receives the response). But in our case it gets bit more complicated because the call also involves getting image data and thus waits for raster thread, which under some circumstances is blocked on platform thread, but fortunately that's not the case here. So the approach from `FlutterKeyboardManager.mm` should work for text input plugin just fine.

**Comment by lukemmtt** [comment]
This issue as a whole seems to have regressed with the changes introduced in iOS 17 Beta. Namely, the autocorrect “highlight rectangle” is now always highlighting the _most recently entered character_, regardless of whether there is an autocorrect pending or not. 

https://github.com/flutter/flutter/assets/1598289/7f22f58d-ff3b-49db-b834-7cf3da6fba72
